servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/AmbSingles.java

```diff
@@ -90,11 +90,13 @@ public void onSubscribe(final Cancellable cancellable) {
 
         @Override
         public void onSuccess(@Nullable final T result) {
+            ignoreCancel();
```

Comment:
There are 2 cases to consider consider:

1. Concurrency between Subscriber and Subscription for one AmbSubscriber object.
2. Concurrency between two different AmbSubscriber when `State` invokes cancel.

for (1) the following sequence is possible:

```
T1 - Subscription.cancel(), DelayedCancellable.GAS currentUpdater reads Cancellable into oldCancellable
T2 - AmbSubscriber.onSuccess(T), DelayedCancellable.ignoreCancel(), state.trySuccess, target.onSuccess
T1 - oldCancellable.cancel()
```

Because they are on independent threads this could happen even with atomic operations (assuming we want to always let Subscriber methods propagate even after cancel) but worth adding a comment to clarify. Can you confirm if this would be a problem for this issue you are observing?

for (2) the following sequence is possible:

```
T1 - AmbSubscriber1.onSuccess, ignoreCancel, state.trySuccess, doneUpdater.GAS, CompositeCancellable.cancel()
T2 - AmbSubscriber2.onSuccess
T1 - AmbSubscriber2.cancel() // cancel already invoked, upstream but already delivered onSuccess
T2 - AmbSubscriber2.ignoreCancel() // cancel already invoked, this won't have any effect
```

Again because each Single can have a different thread for Subscriber and Subscription, and we want to give precedence to the Subscriber, I'm not sure we can prevent this. However worth noting (and add a comment) to confirm this won't cause an issue with your use case.

I suggest clarifying in comments the expectations and constraints:
1. Let Subscriber take precedence - If Subscriber terminates after cancel(), we should propagate as we don't want to suppress work that is already complete and downstream may need to clean up.
2. Best effort enforcement of 2.3 - Subscriber and Subscription can be on different threads, and because of (1) it is possible cancel may happen "after" Subscriber methods are invoked if concurrency is involved in the following scenarios:
 - Concurrency between Subscriber and Subscription for one AmbSubscriber object.
 - Concurrency between two different AmbSubscriber when `State` invokes cancel.

Reply:
Thank you for the detailed description. Imho, both example are legit races that are true for other operators as well. For that type of a race, @bryce-anderson opened #3038. 

Rule 2.3 says:
> `Subscriber.onComplete()` and `Subscriber.onError(Throwable t)` MUST NOT call any methods on the Subscription or the Publisher.

So, in this PR we only try to prevent a case when `onComplete/onError` invokes `cancel()` on the same thread after delivering the terminal. This helps to prevent a cycle.

Reply:
The assumption we currently have in all other places is that if cancel **races** with a terminal on different threads, we should do our best to propagate terminal down, but because we know that we can only do 1 terminal we should pessimistically assume that some other `onError` can win in a chain of operators and the `onSuccess` will be discarded. For this reason, we should always propagate a racy `cancel` up to the originator to make sure it can clean up the state, if necessary.